### PR TITLE
feat: add test overrides for nginx

### PR DESCRIPTION
**Pull request by theseion**

### CODE REVIEW

**Comment by dune73** [comment]
Thanks @theseion. Can we help you get this PR over the line? What is still missing here?

**Comment by theseion** [comment]
Some finishing touches to go-ftw are needed. Need to pick up work on that again. Probably over Christmas.

**Comment by dune73** [comment]
That's a reasonable time frame. Thanks for the info.

Do we need this for v4, or can the label be removed?

**Comment by theseion** [comment]
I think we can ship 4.0 without this.

**Comment by dune73** [comment]
Thank you

**Comment by RedXanadu** [comment]
@theseion Is this still blocked by missing go-ftw features?

**Comment by theseion** [comment]
Yes, I need to implement some things in go-ftw to support multiple engines, the status engine, etc.

**Comment by dune73** [comment]
Keeping my fingers crossed here ... :)

File: tests/regression/nginx-overrides.yaml
Comment (M4tteoP) [inline tests/regression/nginx-overrides.yaml:116]: nit
```suggestion
    reason: "nginx does not strip the dots from the URL (which httpd apparently does), hence ModSecurity sees them and the rule matches"
```
Just to mention it: the same behaviour of not stripping the dots that leads to matching 930110 is happening in Coraza. So maybe the override should live in httpd turning the rule to expect_ids. I will open an Issue/PR.

**Review by fzipi — CHANGES_REQUESTED** [review]
AWESOME! Hopefully we can iterate quickly on this one.

**Review by fzipi — CHANGES_REQUESTED** [review]
AWESOME! Hopefully we can iterate quickly on this one.

**Review by fzipi — CHANGES_REQUESTED** [review]
Once again.

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:20]: Technically, the verb endpoints are no longer necessary. I'm not sure it makes sense to keep them. I've already updated the documentation to reflect that.

Do you think we should keep using the verbs? I'm not sure that there's a benefit.

Comment (fzipi) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:20]: Ugh... yeah. Albedo. 🤦 

Nah, we should remove. But maybe we should clean things in a follow up patch. I don't see why we are writing all the defaults we use in the yaml

Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:20]: So I'll just ignore your suggestions w.r.t. to the changes to `uri`, right?

Yes, there's potential for cleanup. On the other hand, having the defaults makes it immediately clear what the request will look like.

Comment (fzipi) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:20]: > So I'll just ignore your suggestions w.r.t. to the changes to uri, right?

Yes. 👍 👍 👍

**Comment by fzipi** [comment]
@theseion Resolved all comments with change to url. Now all others should be part of the review.

**Review by fzipi — CHANGES_REQUESTED** [review]
There are some tests with `data:` that don't have `method: POST`, and I'm leaving comments on this we might need to create issues to followup after merging this one.

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:141]: I can ignore this one, right?

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:84]: I can ignore this one, right?

File: tests/regression/tests/base_positive_rules.py
Comment (theseion) [reply tests/regression/tests/base_positive_rules.py:20]: We should probably remove the script. In the future, we should have some form of matrix / parameterization for go-ftw instead of hundrds of tests with small variations in the payload. https://github.com/coreruleset/go-ftw/issues/321

**Comment by theseion** [comment]
Thanks for the thorough review @fzipi! 🥇

**Comment by theseion** [comment]
Thanks @M4tteoP!

**Comment by theseion** [comment]
Requires https://github.com/coreruleset/go-ftw/pull/322.

**Comment by theseion** [comment]
@fzipi ping

**Comment by fzipi** [comment]
@theseion looks like we are in conflict now...

**Comment by theseion** [comment]
@fzipi rebased.

**Review by fzipi — COMMENTED** [review]
@theseion So, after the correctness checks, what is your plan here?

I mean, from what I see there is no actual nginx test being run here, right? (e.g. matrix doesn't run for `modsec3-nginx`)

Are we first pushing this change, then adding it to the matrix? Maybe we want also to rename the matrix variable to `platform` instead of `modsec_version`....

**Comment by theseion** [comment]
> I mean, from what I see there is no actual nginx test being run here, right? (e.g. matrix doesn't run for modsec3-nginx)
>
> Are we first pushing this change, then adding it to the matrix? Maybe we want also to rename the matrix variable to platform instead of modsec_version....

I wanted to create a separate PR for both of those.

Once we've merged, I'll also open a PR for submitting the JSON schemas to `schemastore.org` (already prepared).

I haven't yet had a good idea on how to make people aware of the mapping between go-ftw and ftw-tests-schema. I thought about adding a schema version to the spec but that feels a bit clunky.
Another idea would be to make go-ftw fail hard whenever it can't parse a file. I actually changed go-ftw just recently to behave exactly the opposite way and be tolerant of things it doesn't know and simply log warnings. The reason for that was that the globbing very easily leads to inclusion of other files, and when that happens it's very hard for the user to figure out why go-ftw is failing.

Maybe we should couple the versioning of schema and go-ftw. Not a fan of that either though...

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:97]: ping

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920390.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920390.yaml:24]: ping

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920400.yaml
Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920400.yaml:20]: ping

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml
Comment (fzipi) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920170.yaml:97]: Yes

**Review by fzipi — APPROVED** [review]
Somehow I can't see these last two. But they can be forgotten, as they are URL changes.

**Comment by theseion** [comment]
They were hidden (happens when there are too many conversations).

**Pull request merged**
